### PR TITLE
SCUMM: Fix Invalid phrase with GIVE crashes Monkey Island EGA Demo

### PR DESCRIPTION
**Pull request by r41k0u**
Fixes bug [#13196](https://bugs.scummvm.org/ticket/13196)

breathe mint is only meant to be given to actors. when given to non-actor objects in the Monkey Island EGA Demo, it crashed the game. Now it checks if the recipient is an actor before executing the script.

### CODE REVIEW

File: engines/scumm/script_v5.cpp [hunk 1]
```diff
@@ -871,6 +871,14 @@ void ScummEngine_v5::o5_doSentence() {
 
 	int objectA = getVarOrDirectWord(PARAM_2);
 	int objectB = getVarOrDirectWord(PARAM_3);
+
+	//	HACK: breathe mint(object 458) should only be given to actors.
```
Comment (bluegr) [inline engines/scumm/script_v5.cpp:875]:
```suggestion
	//	HACK: breath mint(object 458) should only be given to actors.
```

File: engines/scumm/script_v5.cpp [hunk 2]
```diff
@@ -871,6 +871,14 @@ void ScummEngine_v5::o5_doSentence() {
...
+	//	Giving it to non-actor objects crashes the game.
+	//	The sentence doesn't run a script if the recipient is not an actor.
+	//	Trac #13196
+	if(verb == 3 && ((objectA == 458 && !isValidActor(objectB))))
```
Comment (bluegr) [inline engines/scumm/script_v5.cpp:879]:
```suggestion
	if (verb == 3 && ((objectA == 458 && !isValidActor(objectB))))
```

**Comment by eriktorbjorn** [comment]
Shouldn't it check to make sure which game it's running before applying the workaround?

**Comment by eriktorbjorn** [comment]
I can't even reproduce the bug with the development version of ScummVM. Which particular demo is this, and which objects can you give the breath mint to in order to crash the game? I tried a couple of demos/objects to no avail.

And I'm pretty sure no EGA demo would have GID_MONKEY_VGA as its id. At least it seems unlikely.

**Comment by bluegr** [comment]
> I can't even reproduce the bug with the development version of ScummVM. Which particular demo is this, and which objects can you give the breath mint to in order to crash the game? I tried a couple of demos/objects to no avail.
> 
> And I'm pretty sure no EGA demo would have GID_MONKEY_VGA as its id. At least it seems unlikely.

From the bug description:

> Tested on Windows 10 x64 with the Monkey Island EGA demo from the ScummVM Demos page: "The Secret of Monkey Island (DOS EGA Demo)" (English)

I assume it's this one:

https://downloads.scummvm.org/frs/demos/scumm/monkey1-dos-ega-demo-en.zip

**Comment by antoniou79** [comment]
It is indeed the demo that @bluegr linked. (DOS EGA DEMO). As I noted in the bug report the crash happens if Guybrush attempts to give the breath mint to the pot (even before picking it up), and to other objects, and also if he attempts to give other objects to the breath mint in his inventory (although this reverse case plays ok for the hunk'o'meat).

I haven't tested with the full EGA game, because I don't own it. 
But I think that in the full game Guybrush can (attempt to) give items to other "items" in his inventory (like the Navigator's head) -- or do those count as actors too?

I can reproduce this on Windows 10 build from development branch latest HEAD.

**Comment by eriktorbjorn** [comment]
Thanks, I missed the bug link somehow and didn't see it in the tracker. I was trying to give the breath mint to the pot, not the other way around. (And I was using the pot o' soup, not the portable one.)

So I can reproduce it with the Amiga, English DOS, and German DOS demos. (But not in the Passport to Adventure demo.)

But it looks like it should still check for GID_MONKEY_EGA, not GID_MONKEY_VGA. And, as was previously the case with Loom, there doesn't seem to be any way to distinguish the full game from the demos? So unless the bug also happens in the full game (I don't own that version), or I'm mistaken, something probably has to be done so it can check for _game.features & GF_DEMO.

**Comment by r41k0u** [comment]
While running the game, i logged _game and I got the following:
gameid = "monkey"
variant = "VGA"
preferredTag = "vga"
id = 8
version = 4
heversion = 0
midi = 3087
features = 48
platform = Common::kPlatformAmiga
guioptions = "\003"

id = 8 corresponds to GID_MONKEY_VGA in ScummGameId enum of detection.h ([line 188](https://github.com/scummvm/scummvm/blob/b221812eda028aff879a8153e4495fa3f17cbb53/engines/scumm/detection.h#L197))

I used this version of the game in which the bug is being reproduced.
https://downloads.scummvm.org/frs/demos/scumm/monkey1-amiga-demo-en.zip

Turns out that this bug exists in the VGA demo too and I had the VGA demo instead of the EGA one

**Comment by eriktorbjorn** [comment]
I had forgotten the Amiga demo was VGA. Makes sense, I guess. But I can't reproduce it with the full VGA games (neither floppy nor CD version) and objects are numbered differently when I try to give the pot to the breath mints.

Floppy version: objectA = 333, objectB = 409
CD version: objcetA = 567, objectB = 395

If I didn't mess up when trying.

So I guess it should check for both GID_MONKEY_EGA and GID_MONKEY_VGA, but also it really needs to be able to distinguish the demos from the full games. This is the commit that added the demo flag to the Loom demos: https://github.com/scummvm/scummvm/commit/7df600d6565f7358efd07551d2a25705ae0ac773

**Comment by r41k0u** [comment]
The other demos (EGA english and EGA german) give the following logs:
gameid = "monkey"
variant = "Demo"
preferredTag = "ega"
id = "7"  ([GID_MONKEY_EGA](https://github.com/scummvm/scummvm/blob/b221812eda028aff879a8153e4495fa3f17cbb53/engines/scumm/detection.h#L196))
version = 4
heversion = 0
midi = 15
features = 176
platform = Common::kPlatformDOS
guioptions = "\003\005"

The EGA game has a Demo variant. Need to do that for VGA too.

**Comment by r41k0u** [comment]
Why this bug occurs:
breathe mint is meant to be given to actors. In a playthrough of the EGA Demo, I found out that this is given to a prisoner in jail. before getting the mint the prisoner paces up and down in the cell. But when the give sentence is passed, both guybrush and the prisoner face each other.

During debugging I found that giving breath mint to something calls [o5_faceActor](https://github.com/scummvm/scummvm/blob/547993d966f14907e16130deef7b4075a68924c1/engines/scumm/script_v5.cpp#L1024) twice: Once to make guybrush face the receiving object, then again to make the receiving object face guybrush.

But as in the reported bug's case; pot, red herring and hunk o' meat are not actors, so this results in an error.

Now this error also occurs when pot or red herring is given to breath mint, as they reverse the sentence in their scripts (they call [o5_doSentence](https://github.com/scummvm/scummvm/blob/547993d966f14907e16130deef7b4075a68924c1/engines/scumm/script_v5.cpp#L872) again) with objectA and objectB reversed. This behaviour results in the sentence "give breath mint to pot/red herring" being actually executed, which results in the error.

Giving hunk o' meat to breath mint doesn't trigger this error becuase it doesn't reverse the sentence.

This fix replicates the behaviour of DOSBox as mentioned in the issue. It checks if the object being given is a breath mint and if the recipient is an actor. It does nothing if the recipient is not an actor.

Attaching the "opcodes" logs I made during debugging.
[opcodes.txt](https://github.com/scummvm/scummvm/files/8146653/opcodes.txt)

**Comment by somaen** [comment]
The extra merge-commit that was added in the last push is superfluous, please rebase against master to avoid it.

**Comment by r41k0u** [comment]
conflicts resolved

File: engines/scumm/scumm-md5.h [hunk 3]
```diff
@@ -1,5 +1,9 @@
 /*
+<<<<<<< HEAD
   This file was generated by the md5table tool on Mon Feb  7 12:42:06 2022
+=======
+  This file was generated by the md5table tool on Sat Feb 26 17:16:50 2022
+>>>>>>> 452db9ed48a (SCUMM: seperated AMIGA monkey VGA and VGA Demo variants. Assigned GF_DEMO flags)
```
Comment (bluegr) [inline engines/scumm/scumm-md5.h:6]: The merge conflict markers are still present. Please, resolve them

Comment (r41k0u) [reply engines/scumm/scumm-md5.h]: done

**Comment by bluegr** [comment]
Thank you, this is good to be merged now

I'm squashing the commits into one, as there are a lot of small commits for the same change

**Pull request merged**